x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java

```diff
@@ -1137,6 +1152,209 @@ protected Expression rule(Expression e) {
+                if (opLiteral == null) {
+                    // one of the arithm. operands must be a literal, otherwise the operation wouldn't simplify anything
+                    return true;
+                }
+
+                // Only operations on fixed point literals are supported, since optimizing float point operations can also change the
+                // outcome of the filtering:
+                //    x + 1e18 > 1e18::long will yield different results with a field value in [-2^6, 2^6], optimised vs original;
+                //    x * (1 + 1e-15d) > 1 : same with a field value of (1 - 1e-15d)
+                // so consequently, int fields optimisation requiring FP arithmetic isn't possible either: (x - 1e-15) * (1 + 1e-15) > 1.
+                if (opLiteral.dataType().isRational() || bcLiteral.dataType().isRational()) {
+                    return true;
+                }
+
+                // the Literal will be moved to the right of the comparison, but only if data-compatible with what's there
+                return typesCompatible.apply(bcLiteral.dataType(), opLiteral.dataType()) == false;
+            }
+
+            Expression doSimplify() {
+                // force float point folding for FlP field
+                Literal bcl = operation.dataType().isRational()
+                    ? Literal.of(bcLiteral, ((Number) bcLiteral.value()).doubleValue())
+                    : bcLiteral;
+
+                if (operation instanceof BinaryComparisonInvertible == false) { // should be an assertion?
+                    throw new QlIllegalArgumentException("Unexpected [" + operation.symbol() + "] operation object");
+                }
+                Expression bcRightExpression = ((BinaryComparisonInvertible) operation).binaryComparisonInverse()
+                    .create(bcl.source(), bcl, opRight);
+                bcRightExpression = safeMaybeFold(bcRightExpression);
+                return bcRightExpression == null ? comparison : comparison.replaceChildren(List.of(opLeft, bcRightExpression));
+            }
+
+            static Expression simplify(BinaryComparison comparison, BiFunction<DataType, DataType, Boolean> typesCompatible) {
+                ArithmeticOperation operation = (ArithmeticOperation) comparison.left();
+                // Use symbol comp: SQL operations aren't available in this package (as dependencies)
+                String opSymbol = operation.symbol();
+                // Modulo can't be simplified.
+                if (MOD.symbol().equals(opSymbol)) {
+                    return comparison;
+                }
+                SimplifyOperation simplify = null;
+                if (isMulOrDiv(opSymbol)) {
+                    simplify = new SimplifyMulDiv(comparison);
+                } else if (ADD.symbol().equals(opSymbol) || SUB.symbol().equals(opSymbol)) {
+                    simplify = new SimplifyAddSub(comparison);
+                }
```

Comment:
The SQL operations that this code also acts on aren't available (in the current package) -- there's a comment on the line right before the first in the snippet above. So I don't know how else it can be done.